### DISCUSSION TITLE
Example i2s stream demos all crash ESP32 board

### DISCUSSION THREAD

**Discussion by jeffiel** [pschatzmann/arduino-audio-tools#467] [Q&A] [answered]
Hi - 

I'm trying to get just the basic I2S demos running, but they call crash the board immediately upon startup. No debug messages, just a reboot cycle. File is below, which is just the basic i2s-i2s demo from this repo. Have tried it with other i2s stream demos too, such as CSV, but they all do the same thing. The I2S bus and wiring work - as a sketch directly using the espressif I2S code runs fine. Any tips? Thanks!

Board is an Adafruit ESP32 Feather V2: https://www.adafruit.com/product/5400 

Vanilla demo code (w/ pins changed and Debug mode enabled, but otherwise identical. Sample rate & channels don't change the behavior.)
```
/**
 * @file streams-i2s-i2s.ino
 * @brief Copy audio from I2S to I2S  - I2S uses 1 i2s port
 * @author Phil Schatzmann
 * @copyright GPLv3
 */

#include "AudioTools.h"
 
uint16_t sample_rate=22050;
uint16_t channels = 1;
uint16_t bits_per_sample = 16; // or try with 24 or 32
I2SStream i2s;
StreamCopy copier(i2s, i2s); // copies sound into i2s


// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  // change to Warning to improve the quality
  AudioLogger::instance().begin(Serial, AudioLogger::Debug); 

  // start I2S in
  Serial.println("starting I2S...");
  auto config = i2s.defaultConfig(RXTX_MODE);
  config.sample_rate = sample_rate; 
  config.bits_per_sample = bits_per_sample; 
  config.channels = 2;
  config.i2s_format = I2S_STD_FORMAT;
  config.pin_ws = 15;
  config.pin_bck = 14;
  config.pin_data = 27;
  config.pin_data_rx = 32;
  //config.fixed_mclk = sample_rate * 256;
  // config.pin_mck = 3; // must be 0,1 or 3 - only for ESP_IDF_VERSION_MAJOR >= 4
  i2s.begin(config);

  Serial.println("I2S started...");
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}

```

This is the error that repeats every 10 seconds or so as the board startsup and crashes:

```
rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 271414342, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13160
load:0x40080400,len:3036
entry 0x400805e4

```

**Comment by pschatzmann** [comment, marked as answer]
Never seen this issue before and since you don't get any error messages it is hard to diagnose. 
I would suggest to do the following:

- change the log level both in Arduino and in the sketch to debug, to see where it is crashing
- In the board manager check that you have the latest ESP32 version installed (the current version is 2.0.5).
- Try if you get the same effect by installing the old ESP32 version  1.0.6
- Double check the settings in your Tools menu.
- See https://esp32.com/viewtopic.php?t=25729

**Reply by jeffiel** [reply]
Thanks Phil 

Seems to crash before any debugging makes it out so not much help there.  Latest version of ESP32 confirmed.

I've narrowed it down to the I2SStream declaration. This minimal version crashes:

```
#include "AudioTools.h"
 
uint16_t sample_rate=44100;
uint16_t channels = 2;
uint16_t bits_per_sample = 16; // or try with 24 or 32
I2SStream i2s;


// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  Serial.println("I2S started...");
}

// Arduino loop - copy sound to out 
void loop() {
}
```

and this doesn't crash: 

```
#include "AudioTools.h"
 
uint16_t sample_rate=44100;
uint16_t channels = 2;
uint16_t bits_per_sample = 16; // or try with 24 or 32
// I2SStream i2s;


// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  Serial.println("I2S started...");
}

// Arduino loop - copy sound to out 
void loop() {
}
```
Resulting: 
```
rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 271414342, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1184
load:0x40078000,len:13160
load:0x40080400,len:3036
entry 0x400805e4
[     6][D][esp32-hal-cpu.c:244] setCpuFrequencyMhz(): PLL: 480 / 2 = 240 Mhz, APB: 80000000 Hz
[   230][I][esp32-hal-psram.c:96] psramInit(): PSRAM enabled
I2S started...
```

Only change is the I2SStream declaration. I'm surprised a variable declaration can cause a runtime panic! (But I'm no longer a regular C/C++ practitioner)

**Comment by pschatzmann** [comment]
try to use I2SStream i2s(-1); If this helps change the PIN_I2S_MUTE to -1 in the AudioConfgi.h
I was double checking in the code and there is pinMode(mute_pin, OUTPUT); in the constructor...

But if I remember right, you were writing that the Csv output sketch was failing as well. 
Can you confirm ?

**Reply by jeffiel** [reply]
That fixed it! It runs now. When I was trying CSV output, I was still using I2S as input so was still had I2S definition in there. Wonder why pin 23 (for ESP32 defines) would be problematic, seems like a regular GPIO.... Thank you Phil!

**Reply by pschatzmann** [reply]
Cool, I think I will change the default settings, since this functionality is only rarely used...
Pin23 is a regular GPIO or SPI mosi pin, so this should not really disturb.
